File: app/src/main/java/de/roland_illig/phantomgo/AbstractBoardView.java
package de.roland_illig.phantomgo;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

public abstract class AbstractBoardView extends View {

    private float lastX;
    private float lastY;

    protected abstract Cell getBoard(int x, int y);

    protected abstract int getBoardSize();

    protected void boardMouseClicked(int x, int y) {
    }

    public AbstractBoardView(Context context) {
        super(context);
        init();
    }

    public AbstractBoardView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public AbstractBoardView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent e) {
                lastX = e.getX();
                lastY = e.getY();
                return false;
            }
        });
        setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                int x = screenToBoard(lastX);
                int y = screenToBoard(lastY);
                if (0 <= x && x < getBoardSize() && 0 <= y && y < getBoardSize()) {
                    boardMouseClicked(x, y);
                }
            }
        });
    }

    @Override
    protected void onDraw(Canvas g) {
        super.onDraw(g);

        int bsize = getBoardSize();

        Paint linePaint = solidPaint(0xFF000000);
        linePaint.setStrokeWidth(lineWidth());
        Paint boardPaint = solidPaint(0xFFD48E00);
        Paint blackPaint = solidPaint(0xFF000000);
        Paint whitePaint = solidPaint(0xFFFFFFFF);
        Paint blackTransparentPaint = solidPaint(0x55000000);
        Paint whiteTransparentPaint = solidPaint(0x55FFFFFF);

        g.drawPaint(boardPaint);

        for (int i = 0; i < bsize; i++) {
            int start = boardToScreen(0);
            int end = boardToScreen(bsize - 1);
            int fixed = boardToScreen(i);
            g.drawLine(start, fixed, end, fixed, linePaint);
            g.drawLine(fixed, start, fixed, end, linePaint);
        }

        for (int y = 0; y < bsize; y++) {
            for (int x = 0; x < bsize; x++) {

                Cell cell = getBoard(x, y);
                if (cell.dead || cell.territory != null) {
                    if (cell.dead) {
                        fillCircle(g, x, y, 0.48, cell.color == Player.BLACK ? blackTransparentPaint : whiteTransparentPaint);
                    }
                    if (cell.territory != null) {
                        fillCircle(g, x, y, 0.16, cell.territory == Player.BLACK ? blackPaint : whitePaint);
                    }
                } else if (cell.color != null) {
                    fillCircle(g, x, y, 0.48, cell.color == Player.BLACK ? blackPaint : whitePaint);
                }
            }
        }
    }

    private Paint solidPaint(int color) {
        Paint paint = new Paint();
        paint.setColor(color);
        paint.setStyle(Paint.Style.FILL);
        return paint;
    }

    private void fillCircle(Canvas g, int x, int y, double radius, Paint paint) {
        int top = boardToScreen(y - radius);
        int left = boardToScreen(x - radius);
        int bottom = boardToScreen(y + radius);
        int right = boardToScreen(x + radius);
        g.drawOval(new RectF(left, top, right, bottom), paint);
    }

    private int lineWidth() {
        return Math.max(1, (int) ((boardToScreen(1.0) - boardToScreen(0.0)) / 20.0));
    }

    private int boardToScreen(double bc) {
        int size = Math.min(getWidth(), getHeight());
        return (int) Math.round((size * bc + size) / (getBoardSize() + 1));
    }

    private int screenToBoard(float sc) {
        int size = Math.min(getWidth(), getHeight());
        return (int) Math.round(((double) (sc * (getBoardSize() + 1))) / size - 1);
    }

    protected static class Cell {
        protected final Player color;
        protected final boolean lastMove;
        protected final Player territory;
        protected final boolean dead;

        protected Cell(Player color, boolean lastMove, Player territory, boolean dead) {
            this.color = color;
            this.lastMove = lastMove;
            this.territory = territory;
            this.dead = dead;
        }
    }
}

File: app/src/main/java/de/roland_illig/phantomgo/AbstractBoardView.kt
package de.roland_illig.phantomgo

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.RectF
import android.util.AttributeSet
import android.view.View

abstract class AbstractBoardView : View {

    private var lastX = 0.toFloat()
    private var lastY = 0.toFloat()

    protected abstract val boardSize: Int

    protected abstract fun getBoard(x: Int, y: Int): Cell

    protected open fun boardMouseClicked(x: Int, y: Int) {}

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    private fun init() {
        setOnTouchListener { view, e ->
            lastX = e.x
            lastY = e.y
            false
        }
        setOnClickListener {
            val x = screenToBoard(lastX)
            val y = screenToBoard(lastY)
            if (x in 0 until boardSize && y in 0 until boardSize) {
                boardMouseClicked(x, y)
            }
        }
    }

    override fun onDraw(g: Canvas) {
        super.onDraw(g)

        val bsize = boardSize

        val linePaint = solidPaint(0xFF000000.toInt())
        linePaint.strokeWidth = lineWidth().toFloat()
        val boardPaint = solidPaint(0xFFD48E00.toInt())
        val blackPaint = solidPaint(0xFF000000.toInt())
        val whitePaint = solidPaint(0xFFFFFFFF.toInt())
        val blackTranslucentPaint = solidPaint(0x55000000)
        val whiteTranslucentPaint = solidPaint(0x55FFFFFF)

        g.drawPaint(boardPaint)

        for (i in 0 until bsize) {
            val start = boardToScreen(0.0)
            val end = boardToScreen((bsize - 1).toDouble())
            val fixed = boardToScreen(i.toDouble())
            g.drawLine(start.toFloat(), fixed.toFloat(), end.toFloat(), fixed.toFloat(), linePaint)
            g.drawLine(fixed.toFloat(), start.toFloat(), fixed.toFloat(), end.toFloat(), linePaint)
        }

        for (y in 0 until bsize) {
            for (x in 0 until bsize) {

                val cell = getBoard(x, y)
                if (cell.dead || cell.territory != null) {
                    if (cell.dead) {
                        fillCircle(g, x, y, 0.48, if (cell.color == Player.BLACK) blackTranslucentPaint else whiteTranslucentPaint)
                    }
                    if (cell.territory != null) {
                        fillCircle(g, x, y, 0.16, if (cell.territory == Player.BLACK) blackPaint else whitePaint)
                    }
                } else if (cell.color != null) {
                    fillCircle(g, x, y, 0.48, if (cell.color == Player.BLACK) blackPaint else whitePaint)
                }
            }
        }
    }

    private fun fillCircle(g: Canvas, x: Int, y: Int, radius: Double, paint: Paint) {
        val top = boardToScreen(y - radius)
        val left = boardToScreen(x - radius)
        val bottom = boardToScreen(y + radius)
        val right = boardToScreen(x + radius)
        g.drawOval(RectF(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat()), paint)
    }

    private fun lineWidth(): Int {
        return Math.max(1, ((boardToScreen(1.0) - boardToScreen(0.0)) / 20.0).toInt())
    }

    private fun boardToScreen(bc: Double): Int {
        val size = Math.min(width, height)
        return Math.round((size * bc + size) / (boardSize + 1)).toInt()
    }

    private fun screenToBoard(sc: Float): Int {
        val size = Math.min(width, height)
        return Math.round((sc * (boardSize + 1)).toDouble() / size - 1).toInt()
    }

    protected class Cell(val color: Player?, val territory: Player?, val dead: Boolean)

    companion object {
        private fun solidPaint(color: Int): Paint {
            val paint = Paint()
            paint.color = color
            paint.style = Paint.Style.FILL
            paint.isAntiAlias = true
            return paint
        }
    }
}
